docs/build.rb

```diff
@@ -0,0 +1,1151 @@
+
+    def validate!
+      configured_paths = all_file_paths
+      missing_paths = available_paths.map { |path| path.gsub("#{WIKI_PATH}/", '') } - ignored_paths - existing_docs - configured_paths
+      raise "Unhandled paths #{missing_paths.join(', ')}" if missing_paths.any?
+
+      each do |page|
+        page_keys = page.keys
+        allowed_keys = %i[path new_base_name nav_order title new_path folder children has_children parents]
+        invalid_keys = page_keys - allowed_keys
+        raise "#{page} had invalid keys #{invalid_keys.join(', ')}" if invalid_keys.any?
+      end
+
+      # Ensure unique folder names
+      folder_titles = to_enum.select { |page| page[:folder] }.map { |page| page[:title] }
+      duplicate_folder = folder_titles.tally.select { |_name, count| count > 1 }
+      raise "Duplicate folder titles, will cause issues: #{duplicate_folder}" if duplicate_folder.any?
+
+      # Ensure no folder titles match file titles
+      page_titles = to_enum.reject { |page| page[:folder] }.map { |page| page[:title] }
+      title_collisions = (folder_titles & page_titles).tally
+      raise "Duplicate folder/page titles, will cause issues: #{title_collisions}" if title_collisions.any?
+
+      # Ensure there are no files being migrated to multiple places
+      page_paths = to_enum.reject { |page| page[:path] }.map { |page| page[:title] }
+      duplicate_page_paths = page_paths.tally.select { |_name, count| count > 1 }
+      raise "Duplicate paths, will cause issues: #{duplicate_page_paths}" if duplicate_page_paths.any?
+
+      # Ensure new file paths are only alphanumeric and hyphenated
+      new_paths = to_enum.map { |page| page[:new_path] }
+      invalid_new_paths = new_paths.select { |path| File.basename(path) !~ /^[a-zA-Z0-9_-]*\.md$/ }
+      raise "Only alphanumeric and hyphenated file names required: #{invalid_new_paths}" if invalid_new_paths.any?
+    end
+
+    def available_paths
+      Dir.glob("#{WIKI_PATH}/**/*{.md,.textile}", File::FNM_DOTMATCH)
+    end
+
+    def ignored_paths
+      [
+        '_Sidebar.md',
+        'dev/_Sidebar.md',
+      ]
+    end
+
+    def existing_docs
+      existing_docs = Dir.glob('docs/**/*', File::FNM_DOTMATCH)
+      existing_docs
+    end
+
+    def each(&block)
+      config.each do |parent|
+        recurse(with_metadata(parent), &block)
+      end
+    end
+
+    def all_file_paths
+      to_enum.map { |item| item[:path] }.to_a
+    end
+
+    protected
+
+    # depth first traversal
+    def recurse(parent_with_metadata, &block)
+      block.call(parent_with_metadata)
+      parent_with_metadata[:children].to_a.each do |child|
+        child_with_metadata = with_metadata(child, parents: parent_with_metadata[:parents] + [parent_with_metadata])
+        recurse(child_with_metadata, &block)
+      end
+    end
+
+    def with_metadata(child, parents: [])
+      child = child.clone
+
+      if child[:folder]
+        parent_folders = parents.map { |page| page[:folder] }
```

Comment:
Isn't this operation pointless to do if `parents` is an empty array? Should we not update the code to also check in the `if` condition if `parents` is an empty array?

Reply:
The result is used to create the folder structure on the next line:

```
        parent_folders = parents.map { |page| page[:folder] }
        child[:new_path] = File.join(*parent_folders, child[:folder], 'index.md')
```

The splat operator will handle the case where `parent_folders` is an empty array as expected :+1: